client/dashboard/me/preferences-language/index.tsx

```diff
@@ -0,0 +1,258 @@
+			createSuccessNotice( __( 'Language setting saved.' ), { type: 'snackbar' } );
+			params.delete( 'updated' );
+			window.history.replaceState( {}, '', window.location.pathname );
+		}
+	}, [ createSuccessNotice ] );
+
+	/**
+	 * When we save the language, in case we're using a locale_variant (a language without an official locale)
+	 * the API will return the parent language (example: es-cl will return 'es' in the 'language' field)
+	 * as such, we're overriding the language with the locale_variant, if present, this saves us from checking the data and allows us to
+	 * trust that the 'language' field contains a locale from the languages.
+	 */
+	if ( serverData?.locale_variant ) {
+		serverData.language = serverData.locale_variant;
+	}
+	const data = useMemo(
+		() => ( serverData ? { ...serverData, ...formData } : undefined ),
+		[ serverData, formData ]
+	);
+
+	if ( ! data ) {
+		return null;
+	}
+	// We need the CalypsoLanguage to show the % of translated content in the use_fallback_for_incomplete_languages
+	const selectedLanguage: CalypsoLanguage | undefined = data.language
+		? ( getLanguage( data.language ) as CalypsoLanguage )
+		: undefined;
+
+	const handleSubmit = ( e: React.FormEvent ) => {
+		e.preventDefault();
+		if ( ! formData ) {
+			return;
+		}
+		const mutationData = formData;
+		mutation.mutate( mutationData, {
+			onSuccess: () => {
+				// Ensure the UI picks up the new language by reloading the page.
+				// Add a transient query param so we can show a success notice after reload.
+				const url = new URL( window.location.href );
+				url.searchParams.set( 'updated', 'language' );
+				// Replace to avoid adding an extra history entry.
+				window.location.replace( url.toString() );
+			},
+			onError: ( error ) => {
+				// Prepend previous attempted data back into local edits
+				setFormData( ( current ) => ( { ...mutationData, ...current } ) );
+				createErrorNotice( error.message ?? __( 'Language setting could not be saved.' ), {
+					type: 'snackbar',
+				} );
+			},
+		} );
+	};
+
+	const isDefaultLanguageSelected = !! data.language && isDefaultLocale( data.language );
+	const showIncompleteLocaleControl =
+		! isDefaultLanguageSelected && !! data.language && isTranslatedIncompletely( data.language );
+	const isSaving = mutation.isPending;
+
+	const isDirty =
+		!! formData &&
+		!! serverData &&
+		Object.entries( formData ).some( ( [ key, value ] ) => {
+			return serverData[ key as keyof UserSettingsPreferences ] !== value;
+		} );
+
+	const hasValidLanguage = !! data?.language;
+	const canSubmit = ! isSaving && isDirty && hasValidLanguage;
+	const languageForm = {
+		type: 'regular' as const,
+		labelPosition: 'top' as const,
+		fields: [
+			{
+				id: 'interfaceLanguage',
+				label: __( 'Language' ),
+				children: [
+					'language',
+					'use_fallback_for_incomplete_languages',
+					'enable_translator',
+					'i18n_empathy_mode',
+				],
+			},
+		],
+	};
```

Comment:
```suggestion
	const languageForm: React.ComponentProps<DataForm< UserSettingsPreferences >>['form'] = {
		type: 'regular' as const,
		labelPosition: 'top' as const,
		fields: [
			{
				id: 'interfaceLanguage',
				label: __( 'Language' ),
				children: [
					'language',
					'use_fallback_for_incomplete_languages',
					'enable_translator',
					'i18n_empathy_mode',
				],
			},
		],
	};
```

Try something similar to this (syntax might be slightly wrong). I do this frequently to ensure that I get some type-checking and code-completion at this particular spot rather than down below at the component.

Reply:
added in https://github.com/Automattic/wp-calypso/pull/105311/commits/95a7b6cf75879045084c251a00bf180864655f6c, I've used a shorter notation that I found in another component that, in turn, helped find a type error :) thank you!